File: LongestPalindromicSubstring.cpp
#include <string>
#include <iostream>
using namespace std;

class Solution {
public:
    string longestPalindrome(string s) {
        int n=s.length();
        int r=0,len=0,maxlen=0;
        for(int i=0;i<2*n-1;i++){
            if(i%2==0){
                len=1;
                for(int start=i-2,end=i+2;start>=0&&end<2*n-1;start-=2,end+=2,len+=2){
                    if(s[start]!=s[end])break;
                }
            }else{
                len=0;
                for(int start=i-1,end=i+1;start>=0&&end<2*n-1;start-=2,end+=2,len+=2){
                    if(s[start]!=s[end])break;
                }
            }
            
            if(len>maxlen){
                r=i;
                maxlen=len;
            }
        }

        return s.substr(r-maxlen/2,maxlen);
    }
};

int main(){
    Solution s;
    cout<<s.longestPalindrome("aa")<<endl;

    return 0;
}

File: Longest_Palindromic_Substring.go
package main

import "fmt"

func longestPalindrome(s string) string {
	if len(s) <= 1 {
		return s
	}
	news := make([]rune, 0)
	news = append(news, '*')
	for _, c := range s {
		news = append(news, c)
		news = append(news, '*')
	}
	dp := make([]int, len(news))
	radius := 0
	pivot := 0
	maxLen := 0
	mid := 0
	// print(" ")
	for i := range dp {
		// print(string(news[i]), " ")
		if radius+pivot > i {
			dp[i] = min(pivot+radius-i, dp[pivot-(i-pivot)])
		} else {
			dp[i] = 1
		}

		left, right := i-dp[i], i+dp[i]
		for left >= 0 && right < len(news) && news[left] == news[right] {
			dp[i]++
			left--
			right++
		}

		// fmt.Println(i, dp[i], pivot, radius)

		if dp[i] > maxLen {
			maxLen = dp[i]
			mid = i
		}
		if dp[i]+i > pivot+radius {
			pivot = i
			radius = dp[i]
		}
	}
	// println()
	// fmt.Println(dp)
	return s[(mid-maxLen+1)/2 : (mid-maxLen+1)/2+maxLen-1]
}

func min(a, b int) int {
	if a > b {
		return b
	}
	return a
}

func main() {
	testCases := []string{"abccccdd", "a", "bb", "babad", "babadada"}
	for _, testCase := range testCases {
		fmt.Println(longestPalindrome(testCase))
	}
}
